### PR TITLE
Support multi-gpu in a single process.

### LINKED ISSUE #745: k2.intersect_dense_pruned fails with illegal memory access when used in multi-GPU training
I'm trying to train an MMI-MBR model in a multi-GPU setup in NeMo (the multi-GPU training is powered by pytorch-lightning).

The issue is that `k2.intersect_dense_pruned` works just fine in a single GPU but fails in a two-GPU setup with
```
[F] /k2/k2/csrc/array.h:329:T k2::Array1<T>::operator[](int32_t) const [with T = int; int32_t = int] Check failed: ret == cudaSuccess (700 vs. 0)  Error: an illegal memory access was encountered.
[ Stack-Trace: ]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210518+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2_log.so(k2::internal::GetStackTrace[abi:cxx11](
)+0x58) [0x7efe8266e3c8]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210518+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2context.so(k2::Array1<int>::operator[](int) con
st+0x76a) [0x7efe8278507a]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210518+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2context.so(k2::Renumbering::ComputeOld2New()+0x
1b9) [0x7efe82780c99]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210518+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2context.so(k2::MultiGraphDenseIntersectPruned::
PruneTimeRange(int, int)+0x2c65) [0x7efe829550f5]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210518+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2context.so(std::_Function_handler<void (), k2::
MultiGraphDenseIntersectPruned::Intersect()::{lambda()#1}>::_M_invoke(std::_Any_data const&)+0x197) [0x7efe82955d77]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210518+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2context.so(k2::ThreadPool::ProcessTasks()+0x18b
) [0x7efe82ac765b]
/lib/x86_64-linux-gnu/libstdc++.so.6(+0xd6d84) [0x7eff631b7d84]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x9609) [0x7eff82b2d609]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x43) [0x7eff82a54293]
```
The error is thrown by a process that is on the second GPU.

My software setup:

- Docker image `nvcr.io/nvidia/pytorch:21.03-py3` (python 3.8, cuda 11.2, torch 1.9.0a0)
- k2 from commit `6326266a1a242901449767a00a08b3e8cb4dc6e9`

### PR DESCRIPTION
**Pull request by csukuangfj**
Fixes #745 

The basic idea is to add
```
DeviceGuard guard(context);
```
to every wrapper function.

### CODE REVIEW

File: k2/csrc/eval.h [hunk 1]
```diff
@@ -116,11 +118,12 @@ template <typename ContextPtrType,  // Context*  or ContextPtr ==
                                     // std::shared_ptr<Context>
           typename LambdaT>
 void Eval(ContextPtrType c, int32_t n, LambdaT &lambda) {
+  DeviceGuard guard(c);
   Eval(c->GetCudaStream(), n, lambda);
 }
 
 template <typename LambdaT>
-void EvalDevice(cudaStream_t stream, int32_t n, LambdaT &lambda) {
+static void EvalDevice(cudaStream_t stream, int32_t n, LambdaT &lambda) {
```
Comment (danpovey) [inline k2/csrc/eval.h:126]: Why the `static`?

Comment (csukuangfj) [reply k2/csrc/eval.h:126]: 1. To mark this function being used only inside this file.
2. To reduce the number of global symbols, which I believe can reduce the link time.

**Comment by danpovey** [comment]
Ah OK.  It's just unusual to put `static` in a header.  But I suppose we'd
generally use each lambda only once.
You might want to check the effect on the binary size, if possible.

**Comment by GNroy** [comment]
@csukuangfj Thanks for starting to fix this issue!
I presume that the work is not finished yet, but I could not resist and tried this fix in my setup.
Now I have three processes instead of one stuck in the GPUs doing nothing. But at least none of them failed :)

image

Not sure if it is related to k2, but I've never had a situation before when three processes were spawned instead of two.
I hope this helps you at least somehow.

**Comment by csukuangfj** [comment]
> @csukuangfj Thanks for starting to fix this issue!
> 
> I presume that the work is not finished yet, but I could not resist and tried this fix in my setup.
> 
> Now I have three processes instead of one stuck in the GPUs doing nothing. But at least none of them failed :)
> 
> 
> 
> image
> 
> 
> 
> Not sure if it is related to k2, but I've never had a situation before when three processes were spawned instead of two.
> 
> I hope this helps you at least somehow.

Thank you for for the feedback. I also encountered the same problem as you during the fix. It got stuck in the last kernel call inside the file pytorch_context_test.cu of this pull request, but it was before applying the change with the 'DeviceGuard'.

I will try to finish this pull request today.

File: k2/csrc/eval.h (hunk 1 above)
Comment (csukuangfj) [reply k2/csrc/eval.h]: I have reverted the changes to `eval.h`.

The device guard is put only in the python wrapper.

**Comment by csukuangfj** [comment]
Now all k2 **Python** APIs should support multi-gpu within a single process, except for `k2.intersect_dense_pruned`.

It gets stuck in the following line and I have not found the reason.
https://github.com/k2-fsa/k2/blob/9aefc209f48aac2610f1200415491f6d34bfc626/k2/python/tests/intersect_dense_pruned_test.py#L49

**Comment by csukuangfj** [comment]
After setting the environment variable
```
export K2_SYNC_KERNELS=1
```
It does not get stuck, but throws an error:

```
[F] /root/fangjun/open-source/k2/k2/csrc/eval.h:135:void k2::EvalDevice(cudaStream_t, int32_t, LambdaT&) [with LambdaT = __nv_dl_wrappe
r_t<__nv_dl_tag<void (k2::MultiGraphDenseIntersectPruned::*)(k2::MultiGraphDenseIntersectPruned::FrameInfo*), &k2::MultiGraphDenseInter
sectPruned::SetBackwardProbsFinal, 1>, k2::intersect_pruned_internal::StateInfo*, double>; cudaStream_t = CUstream_st*; int32_t = int]
Check failed: e == cudaSuccess (700 vs. 0)  Error: an illegal memory access was encountered.

```

GDB output
```
(gdb) bt
#0  0x00007ffff1c35d1d in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x00007ffd86615c73 in k2::internal::Logger::~Logger (this=0x7ffd7d442520, __in_chrg=<optimized out>)
    at /root/fangjun/open-source/k2/k2/csrc/log.h:151
#2  0x00007ffd8679d7bc in k2::EvalDevice<__nv_dl_wrapper_t<__nv_dl_tag<void (k2::MultiGraphDenseIntersectPruned::*)(k2::MultiGraphDenseIntersectPruned::FrameInfo*), &k2::MultiGraphDenseIntersectPruned::SetBackwardProbsFinal, 1u>, k2::intersect_pruned_internal::StateInfo*, double> > (stream=0x0, n=1, lambda=...) at /root/fangjun/open-source/k2/k2/csrc/eval.h:134
#3  0x00007ffd86797811 in k2::EvalDevice<std::shared_ptr<k2::Context>, __nv_dl_wrapper_t<__nv_dl_tag<void (k2::MultiGraphDenseIntersectPruned::*)(k2::MultiGraphDenseIntersectPruned::FrameInfo*), &k2::MultiGraphDenseIntersectPruned::SetBackwardProbsFinal, 1u>, k2::intersect_pruned_internal::StateInfo*, double> > (c=std::shared_ptr<k2::Context> (use count 36, weak count 1) = {...}, n=1, lambda=...)
    at /root/fangjun/open-source/k2/k2/csrc/eval.h:143
#4  0x00007ffd8678bb16 in k2::MultiGraphDenseIntersectPruned::SetBackwardProbsFinal (this=0x7fffffffadb0, cur_frame=0x5555f8a74c50)
    at /root/fangjun/open-source/k2/k2/csrc/intersect_dense_pruned.cu:960
#5  0x00007ffd8678d3d0 in k2::MultiGraphDenseIntersectPruned::PruneTimeRange (this=0x7fffffffadb0, begin_t=0, end_t=3)
    at /root/fangjun/open-source/k2/k2/csrc/intersect_dense_pruned.cu:1154
#6  0x00007ffd86788698 in k2::MultiGraphDenseIntersectPruned::BackwardPass (this=0x7fffffffadb0)
    at /root/fangjun/open-source/k2/k2/csrc/intersect_dense_pruned.cu:306
#7  0x00007ffd86788781 in k2::MultiGraphDenseIntersectPruned::BackwardPassStatic (c=0x7fffffffadb0)
    at /root/fangjun/open-source/k2/k2/csrc/intersect_dense_pruned.cu:312
#8  0x00007ffd86787e9b in k2::MultiGraphDenseIntersectPruned::Intersect()::{lambda()#1}::operator()() const (__closure=0x7ffd7d442d80)
    at /root/fangjun/open-source/k2/k2/csrc/intersect_dense_pruned.cu:258
#9  0x00007ffd867a0373 in std::_Function_handler<void (), k2::MultiGraphDenseIntersectPruned::Intersect()::{lambda()#1}>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/include/c++/7/bits/std_function.h:316
#10 0x00007ffd868c9b46 in std::function<void ()>::operator()() const (this=0x7ffd7d442d80)
```

**Comment by csukuangfj** [comment]
Now, k2 should work for multi-gpu training in a single process as well as in multiple processes.

@GNroy 

Could you please try this pull-request again?

**Comment by GNroy** [comment]
I confirm, now it works!
Thank you very much for this fix!

**Comment by GNroy** [comment]
@csukuangfj I'm sorry, I'm afraid I jumped to conclusions.
The training is just got stuck after a few iterations.
I cannot be 100% sure this is because of the intersect_dense_pruned, but it wasn't happening before the fix (on a single GPU).
The GPUs condition in the screenshot remains completely unchanged for a half an hour.
image

**Comment by csukuangfj** [comment]
> @csukuangfj I'm sorry, I'm afraid I jumped to conclusions.
> 
> The training is just got stuck after a few iterations.
> 
> I cannot be 100% sure this is because of the intersect_dense_pruned, but it wasn't happening before the fix (on a single GPU).
> 
> The GPUs condition in the screenshot remains completely unchanged for a half an hour.
> 
> 
> 
> image
> 
> 

Could you provide a demo for reproducing?

**Comment by csukuangfj** [comment]
Can you run 
```
export K2_SYNC_KERNELS=1
```
before running your training script. It may throw some log instead of just hanging forever.

**Comment by GNroy** [comment]
> Could you provide a demo for reproducing?

I'm sorry, it might be hard to provide a good demo.
You can find my code here: https://github.com/GNroy/NeMo/pull/1 but I'm not sure it will be helpful.
What can be helpful is that my `bt` is exactly as in this comment: https://github.com/pytorch/pytorch/issues/45028#issuecomment-695757514 so the issue may be not in k2 itself.

**Comment by GNroy** [comment]
> my bt is exactly as in this comment

Sorry, not exactly, but it is similar:
```
#0  0x00007fcb4bba72d5 in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#1  0x00007fcb4ba28def in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#2  0x00007fcb4ba0e435 in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#3  0x00007fcb4bc80a96 in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#4  0x00007fcb4bc81581 in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#5  0x00007fcb4ba92197 in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#6  0x00007fcb4bb9d4fb in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#7  0x00007fcb4bb9e0d8 in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#8  0x00007fcb4bc8623d in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#9  0x00007fcb4b9f7eac in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#10 0x00007fcb4b9f8585 in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#11 0x00007fcb4b9fa012 in ?? () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#12 0x00007fcb4baa84fe in cuMemcpyDtoHAsync_v2 () from target:/usr/local/cuda/compat/lib.real/libcuda.so
#13 0x00007fcc0a11f4e9 in ?? () from target:/usr/local/cuda/lib64/libcudart.so.11.0
#14 0x00007fcc0a0f99ed in ?? () from target:/usr/local/cuda/lib64/libcudart.so.11.0
#15 0x00007fcc0a13de96 in cudaMemcpyAsync () from target:/usr/local/cuda/lib64/libcudart.so.11.0
#16 0x00007fcbb1f6ab1c in at::native::_local_scalar_dense_cuda(at::Tensor const&)::{lambda()#1}::operator()() const [clone .isra.0] ()
   from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cuda.so
#17 0x00007fcbb1f6be74 in at::native::_local_scalar_dense_cuda(at::Tensor const&) ()
   from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cuda.so
#18 0x00007fcbb3a84ad9 in at::(anonymous namespace)::(anonymous namespace)::wrapper__local_scalar_dense(at::Tensor const&) ()
   from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cuda.so
#19 0x00007fcbb3a84b35 in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<c10::Scalar (at::Tensor const&), &at::(anonymous namespace)::(anonymous namespace)::wrapper__local_scalar_dense>, c10::Scalar, c10::guts::typelist::typelist<at::Tensor const&> >, c10::Scalar (at::Tensor const&)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&) ()
   from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cuda.so
#20 0x00007fcbf59754ec in at::_local_scalar_dense(at::Tensor const&) () from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#21 0x00007fcbf729fc5f in torch::autograd::VariableType::(anonymous namespace)::_local_scalar_dense(at::Tensor const&) ()
   from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#22 0x00007fcbf729fcd5 in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<c10::Scalar (at::Tensor const&), &torch::autograd::VariableType::(anonymous namespace)::_local_scalar_dense>, c10::Scalar, c10::guts::typelist::typelist<at::Tensor const&> >, c10::Scalar (at::Tensor const&)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&) ()
   from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#23 0x00007fcbf59754ec in at::_local_scalar_dense(at::Tensor const&) () from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#24 0x00007fcbf54c1e8b in at::native::item(at::Tensor const&) () from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
```

> Can you run
I did that, thanks!
I'll share the result of it here as something happens.

**Comment by GNroy** [comment]
So something rather unexpected happened.
I ran my training with
```
export K2_SYNC_KERNELS=1
```

and it stuck again near the end of the first epoch but with different error this time:
```
[F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F] [F]
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
/workspace/NeMo/k2/k2/csrc/intersect_dense.cu:846:lambda [](signed int)->void::operator()(signed int)->void
block:[0,0,0], thread: [0,0,0] block:[0,0,0], thread: [1,0,0] block:[0,0,0], thread: [2,0,0] block:[0,0,0], thread: [3,0,0] block:[0,0,0], thread: [4,0,0] block:[0,0,0], thread: [5,0,0] block:[0,0,0], thread: [6,0,0] block:[0,0,0], thread: [7,0,0] block:[0,0,0], thread: [8,0,0] block:[0,0,0], thread: [9,0,0] block:[0,0,0], thread: [10,0,0] block:[0,0,0], thread: [11,0,0] block:[0,0,0], thread: [12,0,0] block:[0,0,0], thread: [13,0,0] block:[0,0,0], thread: [14,0,0] block:[0,0,0], thread: [15,0,0] block:[0,0,0], thread: [16,0,0] block:[0,0,0], thread: [17,0,0] block:[0,0,0], thread: [18,0,0] block:[0,0,0], thread: [19,0,0] block:[0,0,0], thread: [20,0,0] block:[0,0,0], thread: [21,0,0] block:[0,0,0], thread: [22,0,0] block:[0,0,0], thread: [23,0,0] block:[0,0,0], thread: [24,0,0] block:[0,0,0], thread: [25,0,0] block:[0,0,0], thread: [26,0,0] block:[0,0,0], thread: [27,0,0] block:[0,0,0], thread: [28,0,0] block:[0,0,0], thread: [29,0,0] block:[0,0,0], thread: [30,0,0] block:[0,0,0], thread: [31,0,0]
Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed: Check failed:
tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0tot_score_end == tot_score_start || fabs(tot_score_end - tot_score_start) < 1.0
nannannannannannannannannannannannannannannannannannannannannannannannannannannannannannannannan vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs  vs nannannannannannannannannannannannannannannannannannannannannannannannannannannannannannannannan

[F] /workspace/NeMo/k2/k2/csrc/array.h:329:T k2::Array1<T>::operator[](int32_t) const [with T = int; int32_t = int] Check failed: ret == cudaSuccess (710 vs. 0)  Error: device-side assert triggered. 


[ Stack-Trace: ]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210528+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2_log.so(k2::internal::GetStackTrace[abi:cxx11]()+0x58) [0x7fdcab3d53c8]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210528+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2context.so(k2::Array1<int>::operator[](int) const+0x76a) [0x7fdc94701d9a]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210528+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2context.so(k2::Renumbering::ComputeOld2New()+0x1b9) [0x7fdc946fd9b9]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210528+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2context.so(k2::Renumbering::ComputeNew2Old()+0x7a0) [0x7fdc946fefa0]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210528+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2context.so(k2::MultiGraphDenseIntersect::FormatOutput(k2::Array1<int>*, k2::Array1<int>*)+0x73c) [0x7fdc948aef6c]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210528+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/libk2context.so(k2::IntersectDense(k2::Ragged<k2::Arc>&, k2::DenseFsaVec&, k2::Array1<int> const*, float, k2::Ragged<k2::Arc>*, k2::Array1<int>*, k2::Array1<int>*)+0x4d2) [0x7fdc9489b1f2]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210528+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/_k2.cpython-38-x86_64-linux-gnu.so(+0x7bbaf) [0x7fdcab4d9baf]
/opt/conda/lib/python3.8/site-packages/k2-0.3.4.dev20210528+cuda11.2.torch1.9.0a0-py3.8-linux-x86_64.egg/_k2.cpython-38-x86_64-linux-gnu.so(+0x2eaa5) [0x7fdcab48caa5]
```

Now the problem is with k2.intersect_dense of the numerator lattice computing.
Nevertheless, it seems like the issues are not directly related to the topic of this PR.
Thanks!

**Comment by danpovey** [comment]
It looks like there may have been divergence in your model so there were
nan's or inf's in the output.

**Comment by GNroy** [comment]
@danpovey 
Thanks for the suggestion!
So far I did several launches of MMI and every one of them failed after a few epochs with the error above on one of the GPUs.
It doesn't seem like a divergence issue (at least WER and the loss decreased expectedly at every training attempt).
I'll open a separate issue as it is most likely unrelated to this PR.

**Comment by danpovey** [comment]
I recommend to test your nnet output for NaN's and inf's.  The fact that
the message says nan indicates a bad number is getting in there somewhere.
What is your nnet topology?

**Comment by GNroy** [comment]
I'm trying to train a Citrinet from the Nemo toolkit with an MMI (with or without MBR) loss and the standard CTC topology (not modified or pruned).
Thanks, I'll try checking the logsoftmax outputs for nan's and -inf's before using the intersection.

**Comment by danpovey** [comment]
Did you use the same special optimizer they mention here
https://arxiv.org/pdf/2104.01721.pdf

"To train Citrinet models we used the NovoGrad optimizer [23] with learning
rate (LR) of 0.05, β1 = 0.8, β2 = 0.25 and weight decay of 0.001. We used a
cosine annealing LR policy with 1K step linear warmup.".

.. I'm a little surprised that for a convolution network with plenty of
batch-norm, you'd need such a finely tuned optimization approach.

**Comment by jtrmal** [comment]
OT -- but what is that gpu process monitor?
y.

**Comment by GNroy** [comment]
@danpovey 
I've managed to isolate the NaN issue.
You were right, it was a divergence, though a weird one.
It happened only when the nnet was trained in float16. After some number of epochs the model started constantly emitting NaNs at logsoftmax.
This issue did not occur when when training with float32.
IMO this PR can be merged since it has nothing to do with the issue.
And I'll continue to investigate this behaviour since I'm not ready to give up on float16 training just yet.

> Did you use the same special optimizer they mention here

Yes, with those exactly parameters. In my last experiments, the convergence speed and curvature are nearly the same for MMI as for CTC.

> .. I'm a little surprised that for a convolution network with plenty of batch-norm, you'd need such a finely tuned optimization approach.

Can't say anything here because I wasn't part of the team back when they developed it.

Thanks a lot for your help! It was very useful.

@jtrmal 

> OT -- but what is that gpu process monitor?

Do you mean on screenshots? It is `nvtop` (NVidia TOP). I find it quite pretty.

**Comment by csukuangfj** [comment]
I am merging it as the problem is not happening.

**Pull request merged**